File: backend/app/schemas/for_review.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlmodel import SQLModel

from app.enums import ReasonCategory, ReviewEntityType
from app.schemas.job_order import JobItemPublic, JobOrderPublic


class ForReviewPublic(SQLModel):
    id: uuid.UUID
    entity_type: ReviewEntityType
    entity_id: uuid.UUID
    entity_reference: str
    reason: str
    reason_category: ReasonCategory
    resolution_note: str | None
    created_at: datetime
    created_by_name: str
    resolved_at: datetime | None
    resolved_by_id: uuid.UUID | None
    resolved_by_name: str | None = None
    
    
class PossibleJobOrder(SQLModel):
    id: uuid.UUID
    jo_number: int
    job_items: list[JobItemPublic]
    customer_name: str | None
    date_received: datetime
    total_due: Decimal
    total_paid: Decimal
    remaining_balance: Decimal
    match_score: int
    match_reasons: list[str]


class UnlinkedPaymentWithJobMatch(SQLModel):
    id: uuid.UUID
    date_received: datetime
    reference_number: str | None
    amount: Decimal
    customer_name: str | None
    description: str | None
    account_id: uuid.UUID
    account_name: str | None


class UnlinkedPaymentReviewData(UnlinkedPaymentWithJobMatch):
    possible_matches: list[PossibleJobOrder]
    
    
class ForReviewDetails(ForReviewPublic):
    entity: UnlinkedPaymentReviewData | JobOrderPublic

File: frontend/src/types/forReview.ts
import type { JobOrder, JobItem } from "./jobOrder"

export type ReviewEntityType =
    | 'Job Order'
    | 'Payment'
    | 'Expense'
    | 'Misc Sale'

export type ReasonCategory =
    | 'Pricing Discrepancy'
    | 'Missing Data'
    | 'Status Issue'
    | 'Needs Verification'

export interface ForReview {
    id: string
    entity_type: ReviewEntityType
    entity_id: string
    entity_reference: string
    reason: string
    reason_category: ReasonCategory
    created_at: Date
    created_by_name: string
    resolved_at?: Date
    resolved_by_id?: string | null
    resolved_by_name?: string | null
}

export interface PossibleMatches {
    id: string
    jo_number: number
    job_items: JobItem[]
    customer_name?: string
    date_received: Date
    total_due: number
    total_paid: number
    remaining_balance: number
    match_score: number
    match_reasons: string[]
}

export interface UnlinkedPayment {
    id: string
    date_received: Date
    reference_number?: string | null
    amount: number
    customer_name?: string | null
    description?: string | null
    account_id: string
    account_name?: string | null
    possible_matches: PossibleMatches[]
}

export interface PaymentForReview extends ForReview {
    entity: UnlinkedPayment
}